File: capture-tui-go/internal/parser/extractors.go
package parser

import (
	"regexp"
	"strings"
)

// TaskExtractor extracts tasks from content
type TaskExtractor struct {
	patterns []*regexp.Regexp
}

// NewTaskExtractor creates a new task extractor
func NewTaskExtractor() *TaskExtractor {
	patterns := []*regexp.Regexp{
		regexp.MustCompile(`(?m)^- \[ \] (.+)$`),           // Markdown todo
		regexp.MustCompile(`(?i)TODO[:：]\s*(.+)$`),         // TODO: task
		regexp.MustCompile(`(?i)任务[:：]\s*(.+)$`),          // 任务: task
		regexp.MustCompile(`(?i)ACTION[:：]\s*(.+)$`),        // ACTION: task
	}

	return &TaskExtractor{patterns: patterns}
}

// Extract extracts tasks from content
func (te *TaskExtractor) Extract(content string) []string {
	seen := make(map[string]bool)
	var tasks []string

	for _, pattern := range te.patterns {
		matches := pattern.FindAllStringSubmatch(content, -1)
		for _, match := range matches {
			if len(match) > 1 {
				task := strings.TrimSpace(match[1])
				if task != "" && !seen[task] {
					seen[task] = true
					tasks = append(tasks, task)
				}
			}
		}
	}

	return tasks
}

// TagExtractor extracts tags from content
type TagExtractor struct {
	pattern *regexp.Regexp
}

// NewTagExtractor creates a new tag extractor
func NewTagExtractor() *TagExtractor {
	return &TagExtractor{
		pattern: regexp.MustCompile(`#(\w+)`),
	}
}

// Extract extracts tags from content
func (te *TagExtractor) Extract(content string) []string {
	matches := te.pattern.FindAllStringSubmatch(content, -1)
	seen := make(map[string]bool)
	var tags []string

	for _, match := range matches {
		if len(match) > 1 {
			tag := strings.ToLower(strings.TrimSpace(match[1]))
			if tag != "" && !seen[tag] {
				seen[tag] = true
				tags = append(tags, tag)
			}
		}
	}

	return tags
}

// PriorityExtractor extracts priority from content
type PriorityExtractor struct {
	patterns map[string][]*regexp.Regexp
}

// NewPriorityExtractor creates a new priority extractor
func NewPriorityExtractor() *PriorityExtractor {
	return &PriorityExtractor{
		patterns: map[string][]*regexp.Regexp{
			"P0": {
				regexp.MustCompile(`\bP0\b`),
				regexp.MustCompile(`【紧急】`),
				regexp.MustCompile(`【高优先级】`),
				regexp.MustCompile(`(?i)\burgent\b`),
				regexp.MustCompile(`(?i)\bcritical\b`),
				regexp.MustCompile(`(?i)\bASAP\b`),
			},
			"P1": {
				regexp.MustCompile(`\bP1\b`),
				regexp.MustCompile(`【重要】`),
				regexp.MustCompile(`(?i)\bimportant\b`),
				regexp.MustCompile(`(?i)high priority`),
			},
			"P2": {
				regexp.MustCompile(`\bP2\b`),
				regexp.MustCompile(`【一般】`),
				regexp.MustCompile(`(?i)\bnormal\b`),
				regexp.MustCompile(`(?i)low priority`),
			},
		},
	}
}

// Extract extracts priority from content
func (pe *PriorityExtractor) Extract(content string) string {
	// Check P0 first, then P1, then P2
	for _, priority := range []string{"P0", "P1", "P2"} {
		for _, pattern := range pe.patterns[priority] {
			if pattern.MatchString(content) {
				return priority
			}
		}
	}
	return "P2" // Default priority
}

// TitleExtractor extracts title from content
type TitleExtractor struct {
	maxLength int
}

// NewTitleExtractor creates a new title extractor
func NewTitleExtractor() *TitleExtractor {
	return &TitleExtractor{maxLength: 100}
}

// Extract extracts title from content
func (te *TitleExtractor) Extract(content string) string {
	lines := strings.Split(content, "\n")
	
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		// Check for markdown heading
		if strings.HasPrefix(line, "#") {
			line = strings.TrimLeft(line, "# ")
			return te.truncate(line)
		}

		// Use first non-empty line
		return te.truncate(line)
	}

	return "Untitled"
}

// truncate truncates text to max length
func (te *TitleExtractor) truncate(text string) string {
	if len(text) > te.maxLength {
		return text[:te.maxLength]
	}
	return text
}

// SetMaxLength sets the maximum title length
func (te *TitleExtractor) SetMaxLength(length int) {
	te.maxLength = length
}

File: capture/capture_tui/parser/extractors.py
"""内容提取器"""

import re
from typing import List


class TaskExtractor:
    """任务提取器"""
    
    PATTERNS = [
        r'- \[ \] (.+)',          # Markdown 待办
        r'TODO[:：]\s*(.+)',      # TODO: 任务
        r'任务[:：]\s*(.+)',      # 任务: 任务
        r'ACTION[:：]\s*(.+)',     # ACTION: 动作
        r'(?:需要|应该|要)\s*(.+?)(?:，|。|$)',  # 需要/应该/要 ...
    ]
    
    def extract(self, content: str) -> List[str]:
        """提取任务列表"""
        tasks = []
        for pattern in self.PATTERNS:
            matches = re.findall(pattern, content, re.IGNORECASE | re.MULTILINE)
            tasks.extend(matches)
        
        # 去重并清理
        seen = set()
        cleaned = []
        for task in tasks:
            task = task.strip()
            if task and task not in seen:
                seen.add(task)
                cleaned.append(task)
        
        return cleaned


class TagExtractor:
    """标签提取器"""
    
    PATTERN = r'#(\w+)'
    
    # 常见的标签词
    COMMON_TAGS = [
        'feature', 'bug', 'improvement', 'docs', 'refactor',
        'ui', 'api', 'backend', 'frontend', 'test',
        'urgent', 'important', 'later', 'idea'
    ]
    
    def extract(self, content: str) -> List[str]:
        """提取标签"""
        # 提取 #tag 格式
        tags = re.findall(self.PATTERN, content)
        
        # 转换为小写并去重
        tags = list(set(t.lower() for t in tags))
        
        return tags


class PriorityExtractor:
    """优先级提取器"""
    
    PATTERNS = {
        'P0': [
            r'P0',
            r'【紧急】',
            r'【高优先级】',
            r'urgent',
            r'critical',
            r'ASAP',
        ],
        'P1': [
            r'P1',
            r'【重要】',
            r'important',
            r'high priority',
        ],
        'P2': [
            r'P2',
            r'【一般】',
            r'normal',
            r'low priority',
        ]
    }
    
    def extract(self, content: str) -> str:
        """提取优先级"""
        # 中文模式不需要转小写
        for priority, patterns in self.PATTERNS.items():
            for pattern in patterns:
                # 检查原始内容（用于中文匹配）
                if re.search(pattern, content):
                    return priority
                # 检查小写内容（用于英文匹配）
                if re.search(pattern, content, re.IGNORECASE):
                    return priority
        
        return 'P2'  # 默认优先级


class TitleExtractor:
    """标题提取器"""
    
    def extract(self, content: str, max_length: int = 100) -> str:
        """提取标题"""
        lines = content.strip().split('\n')
        
        for line in lines:
            line = line.strip()
            if not line:
                continue
            
            # 如果是 Markdown 标题
            if line.startswith('#'):
                return line.lstrip('#').strip()[:max_length]
            
            # 普通文本取第一行
            return line[:max_length]
        
        return "Untitled"
